src/webui/api/torrentscontroller.cpp

```diff
@@ -1523,3 +1582,79 @@ void TorrentsController::setSSLParametersAction()
 
     torrent->setSSLParameters(sslParams);
 }
+
+void TorrentsController::metadataAction()
+{
+    const QString source {params()[u"source"_s]};
+
+    if (Net::DownloadManager::hasSupportedScheme(source))
+        throw APIError(APIErrorType::BadParams, tr("Source must be magnet URI or info hash"));
+
+    if (const auto parseResult = BitTorrent::TorrentDescriptor::parse(source))
+    {
+        auto *session = BitTorrent::Session::instance();
+        const BitTorrent::TorrentDescriptor &torrentDescr = parseResult.value();
+        const BitTorrent::InfoHash &infoHash = torrentDescr.infoHash();
+
+        const BitTorrent::Torrent *const torrent = BitTorrent::Session::instance()->findTorrent(infoHash);
+        // torrent has already been added and has metadata
+        if (torrent && torrent->info().isValid())
+        {
+            // the TorrentInfo returned by Torrent::info() contains an empty list of trackers
+            // we must fetch the trackers directly via Torrent::trackers()
+            QVector<BitTorrent::TrackerEntry> trackers;
+            for (const BitTorrent::TrackerEntryStatus &tracker : torrent->trackers())
+                trackers << BitTorrent::TrackerEntry
+                {
+                    .url = tracker.url,
+                    .tier = tracker.tier
+                };
+
+            setResult(serializeTorrentMetadata(torrent->info(), trackers));
+            return;
+        }
+
+        const auto iter = m_torrentMetadata.find(infoHash);
+
+        // metadata has already been downloaded
+        if (iter != m_torrentMetadata.end() && iter.value().isValid())
+        {
+            const BitTorrent::TorrentInfo &metadata = iter.value();
+            setResult(serializeTorrentMetadata(metadata, metadata.trackers()));
+            return;
+        }
+
+        if (!session->isKnownTorrent(infoHash) && iter == m_torrentMetadata.end())
+        {
+            qDebug("Fetching metadata for %s", qUtf8Printable(infoHash.toTorrentID().toString()));
+            if (!session->downloadMetadata(torrentDescr))
+                throw APIError(APIErrorType::BadParams, tr("Unable to download metadata"));
+
+            m_torrentMetadata.insert(infoHash, BitTorrent::TorrentInfo{});
```

Comment:
```suggestion
            m_torrentMetadata.insert(infoHash, BitTorrent::TorrentInfo());
```

Reply:
It can be even shorter:
`m_torrentMetadata.insert(infoHash, {});`